python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -241,20 +249,33 @@ def _update(self):
             return
 
         self.last_update_time = now
-        self.update_worker_list()
+
+        # Query node provider to update workers, all_workers, and all_nodes.
+        self.update_node_lists()
 
         # Remove from LoadMetrics the ips unknown to the NodeProvider.
-        self.load_metrics.prune_active_ips([
-            self.provider.internal_ip(node_id) for node_id in self.all_workers
-        ])
+        self.load_metrics.prune_active_ips(
+            [self.provider.internal_ip(node_id) for node_id in self.all_nodes])
+
```

Comment:
why self.all_nodes instead of self.all_workers?

Reply:
because the head node is not a worker node

Reply:
Oh, I see.

Reply:
Misinterpreted the question.

Because previously the load metrics method secretly hacked in the head node ip, leading to weird bugs when writing new node providers.
Now load metrics does not have the head ip as an instance variable and the ip is passed in here.

Reply:
But won't it always be active?

Reply:
Yes. Setting head ip active in load metrics maybe saves us the redundancy of passing it in each time, but it makes the code more brittle. There have been bugs from setting the head ip differently when initializing load metrics than when reading from node provider.